tests/31sync/09archived.pl

```diff
@@ -56,16 +55,67 @@
       })->then( sub {
          my ( $body ) = @_;
 
+         log_if_fail "sync result", $body;
+
          my $room = $body->{rooms}{leave}{$room_id};
          assert_json_keys( $room, qw( timeline state ));
 
          @{ $room->{state}{events} } == 0
             or die "Expected no state events";
 
+         # we should see our own leave event
+         assert_eq(scalar @{ $room->{timeline}{events} }, 1,
+                   "timeline events");
+         my $ev = $room->{timeline}{events}[0];
+         assert_eq($ev->{type}, "m.room.member", "event type");
+         assert_eq($ev->{content}{membership}, "leave", "membership");
+
          Future->done(1);
       });
    };
 
+test "We should see our own leave event, even if history_visibility is " .
+    "restricted (SYN-662)",
+   requires => [ local_user_fixture( with_events => 0 ),
+                 qw( can_sync ) ],
+
+   check => sub {
+      my ( $user ) = @_;
+
+      my ( $filter_id, $room_id );
+
+     matrix_create_filter( $user,
+         { room => { include_leave => JSON::true } }
+     )->then( sub {
+         ( $filter_id ) = @_;
+
+         matrix_create_room( $user );
+      })->then( sub {
+         ( $room_id ) = @_;
+         matrix_set_room_history_visibility( $user, $room_id, "joined" );
+      })->then( sub {
+         matrix_sync( $user, filter => $filter_id );
+      })->then( sub {
+         matrix_leave_room( $user, $room_id );
+      })->then( sub {
+         matrix_sync_again( $user, filter => $filter_id );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         log_if_fail "sync result", $body;
+
+         my $room = $body->{rooms}{leave}{$room_id};
+         assert_json_keys( $room, qw( timeline state ));
+
+         @{ $room->{state}{events} } == 0
+            or die "Expected no state events";
+
+         assert_eq(scalar @{ $room->{timeline}{events} }, 1,
+                   "timeline events");
```

Comment:
Could you pick just one style of testing these; rather than using two different ones for different fields in the same test?